### PR TITLE
Fix Oracle pagination and sorting/grouping problem: add ‘order by’/'group by' of the subquery  to its outer statement if the subquery selects oracle rownum

### LINKED ISSUE #14334: Oracle pagination and sorting doesn' t work properly
## Question

What happens?
I'm using sharding-jdbc springboot, trying to perform pagination (with PageHelper) with 2 sharding tables (in Oracle).
The result sets seem like containing all the result I need, but they have not been sorted.
So I wonder if i configure it wrong or what?
I have searched for it for a long time, but I never found a proper solution.

Here's some configuration and the test code:
**Maven Dependency:**
image

**Configuration:**
image

**Sharding Algorithm:**
Plz ignore it for now. It seems to have nothing to do with this issue.

**Test Code:**
image

image

**Mybatis Mapper:**
image

**Here's the result sets:**
image

**The SQL Log:**
image

What it appears?
And above all, we can see that it scanned through 2 tables, nfs_push_sent_message and nfs_push_sent_message_2021, and the sql did have the "order by STATUS_TIME desc" phase clearly. But still, the "order column" in **result sets were not ordered**.

If you need any further information, please contact me at any time. Thank for for taking a look at the question.

### PR DESCRIPTION
**Pull request by xinglijun1973**
Fixes #14334, #13814:
Oracle pagination and sorting doesn' t work properly #14334,
Using sum function does not return data #13814.

Changes proposed in this pull request:
-if use oracle's rownun alias in where, it means use the "order by"/"group by" exists in  the rownum's from sub query, 
ex: `select * from (select t.*, rownum r from (select * from a order by a.name)t ) where r<10`
means:
`select * from (select t.*, rownum r from (select * from a order by a.name)t ) where r<10 orde by name`, so in `SelectStatementContext`
-first, resolve ‘rownum alias’ and ‘the indeed oracelSelectStatement’ relation 
-second, using the relation above, if use rownum alias in where, update 'order by'/'group by' context.

### CODE REVIEW

**Review by strongduanmu — CHANGES_REQUESTED** [review]
Hi @xinglijun1973, thank you for your contribution. Can you fix ci error first?

**Comment by xinglijun1973** [comment]
had fixed



---Original---
From: "Zhengqiang ***@***.***&gt;
Date: Sat, Aug 20, 2022 21:52 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [apache/shardingsphere] 14334 (PR #20257)




 
@strongduanmu requested changes on this pull request.
 
Hi @xinglijun1973, thank you for your contribution. Can you fix ci error first?

**Comment by strongduanmu** [comment]
Hi @xinglijun1973, can you modify the pr title for friendly search?

**Comment by xinglijun1973** [comment]
@strongduanmu done

**Comment by strongduanmu** [comment]
> @strongduanmu done

Good job, I will review this pr again.

File: infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/ProjectionUtil.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * projection util. 
+ */
+public class ProjectionUtil {
+    
+    /**
+     * expand projection, such as ShorthandProjection.
+     * @param projection the projection to expand
+     * @return expanded projections
+     */
+    public static Collection<Projection> expandProjection(final Projection projection) {
```
Comment (strongduanmu) [inline infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/ProjectionUtil.java:36]: Can we get expand projections by projection context?

File: infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/dml/OracleSelectStatementContextDialect.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.column.ColumnSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.BinaryOperationExpression;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.ExpressionSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.dml.predicate.WhereSegment;
+import org.apache.shardingsphere.sql.parser.sql.dialect.statement.oracle.dml.OracleSelectStatement;
+
+/**
+ * Oracle Select SQL statement context dialect.
+ */
+public final class OracleSelectStatementContextDialect implements SelectStatementContextDialect {
```
Comment (strongduanmu) [inline infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/dml/OracleSelectStatementContextDialect.java:35]: Can logic from OracleSelectStatementContextDialect be merged into SelectStatementContext ?

File: infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/dml/SelectStatementContextDialect.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ * Select SQL statement context dialect.
+ */
+public interface SelectStatementContextDialect {
+    
+    /**
+     * update GroupbyContext.
+     * @param groupByContext groupByContext
+     */
+    default void updateGroupbyContext(final GroupByContext groupByContext) {
+
```
Comment (strongduanmu) [inline infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/dml/SelectStatementContextDialect.java:34]: Please remove useless blank line.

File: infra/binder/src/test/java/org/apache/shardingsphere/infra/binder/statement/impl/SelectStatementContextTest.java [hunk 4]
```diff
@@ -602,4 +604,33 @@ private ProjectionSegment getColumnProjectionSegmentWithoutOwner(final boolean h
...
+        rownumProjectionSegment.setAlias(new AliasSegment(0, 0, new IdentifierValue("pagerow_")));
+        innerProjections.getProjections().add(rownumProjectionSegment);
+        ColumnSegment cityColumnSegment = new ColumnSegment(0, 0, new IdentifierValue("city"));
+        innerProjections.getProjections().add(new ColumnProjectionSegment(cityColumnSegment));
+        innerProjections.getProjections().add(new AggregationProjectionSegment(0, 0, AggregationType.COUNT, "(1)"));
+        OracleSelectStatement innerSelectStatement = new OracleSelectStatement();
+        innerSelectStatement.setProjections(innerProjections);
+        innerSelectStatement.setFrom(new SimpleTableSegment(new TableNameSegment(0, 0, new IdentifierValue("t_user"))));
+        innerSelectStatement.setGroupBy(new GroupBySegment(0, 0, Arrays.asList(new ColumnOrderByItemSegment(cityColumnSegment, OrderDirection.ASC))));
+
```
Comment (strongduanmu) [inline infra/binder/src/test/java/org/apache/shardingsphere/infra/binder/statement/impl/SelectStatementContextTest.java:622]: Please remove this useless black line.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/dml/OracleSelectStatementContextDialect.java (hunk 2 above)
Comment (xinglijun1973) [reply infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/dml/OracleSelectStatementContextDialect.java]: had it merged

File: infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/engine/ProjectionsContextEngine.java [hunk 5]
```diff
@@ -115,32 +114,37 @@ private boolean containsProjection(final OrderByItemSegment orderByItem, final C
...
         return false;
     }
     
-    private Collection<ColumnProjection> getColumnProjections(final Projection projection) {
-        Collection<ColumnProjection> result = new LinkedList<>();
-        if (projection instanceof ColumnProjection) {
-            result.add((ColumnProjection) projection);
-        }
+    /**
+     * expand projection, such as ShorthandProjection.
```
Comment (strongduanmu) [inline infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/engine/ProjectionsContextEngine.java:131]: Please keep the first letter uppercase in javadoc, and add a new line between javadoc and param doc.

File: infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/dml/SelectStatementContext.java [hunk 6]
```diff
@@ -111,11 +113,26 @@ public SelectStatementContext(final ShardingSphereMetaData metaData, final List<
         ColumnExtractor.extractColumnSegments(columnSegments, whereSegments);
         subqueryContexts = createSubqueryContexts(metaData, params, defaultDatabaseName);
         tablesContext = new TablesContext(getAllTableSegments(), subqueryContexts, getDatabaseType());
-        String databaseName = tablesContext.getDatabaseName().orElse(defaultDatabaseName);
+        Optional<SelectStatementContext> rownumSelectContext;
+        if (sqlStatement instanceof OracleSelectStatement && sqlStatement.getWhere().isPresent()) {
```
Comment (strongduanmu) [inline infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/dml/SelectStatementContext.java:117]: Can we extract if condition to a private method.

File: infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/dml/SelectStatementContext.java [hunk 7]
```diff
@@ -130,6 +147,42 @@ private Map<Integer, SelectStatementContext> createSubqueryContexts(final Shardi
...
+        WhereSegment where = getSqlStatement().getWhere().get();
+        Optional<OracleSelectStatement> rownumSelectStatement = Optional.empty();
+        if (where.getExpr() instanceof BinaryOperationExpression) {
+            rownumSelectStatement = findRownumSelectStatement((BinaryOperationExpression) where.getExpr(), (OracleSelectStatement) getSqlStatement());
+        }
+        Optional<SelectStatementContext> rownumSelectContext = Optional.empty();
+        if (rownumSelectStatement.isPresent()) {
+            rownumSelectContext = findContextFromSubQueryContext(subqueryContexts.values(), rownumSelectStatement.get());
+        }
+        return rownumSelectContext;
```
Comment (strongduanmu) [inline infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/dml/SelectStatementContext.java:160]: Please rename rownumSelectContext to result.

File: infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/dml/SelectStatementContext.java [hunk 8]
```diff
@@ -130,6 +147,42 @@ private Map<Integer, SelectStatementContext> createSubqueryContexts(final Shardi
...
+    
+    private Optional<SelectStatementContext> findContextFromSubQueryContext(final Collection<SelectStatementContext> subQueryContexts, final OracleSelectStatement selectStatement) {
+        Optional<SelectStatementContext> statementContext = Optional.empty();
+        for (SelectStatementContext subSelectContext : subQueryContexts) {
+            if (subSelectContext.getSqlStatement() == selectStatement) {
+                statementContext = Optional.of(subSelectContext);
+                break;
+            }
+        }
+        return statementContext;
```
Comment (strongduanmu) [inline infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/dml/SelectStatementContext.java:183]: Please rename rownumSelectContext to result.

File: sql-parser/dialect/oracle/src/main/java/org/apache/shardingsphere/sql/parser/oracle/visitor/statement/impl/OracleDMLStatementSQLVisitor.java [hunk 9]
```diff
@@ -512,6 +515,31 @@ public ASTNode visitQueryBlock(final QueryBlockContext ctx) {
         if (null != ctx.modelClause()) {
             result.setModelSegment((ModelSegment) visit(ctx.modelClause()));
         }
+        Map<String, OracleSelectStatement> rowNumSelect = getRowNumSelect(result);
+        if (rowNumSelect.size() > 0) {
+            result.setRowNumSelect(rowNumSelect);
```
Comment (strongduanmu) [inline sql-parser/dialect/oracle/src/main/java/org/apache/shardingsphere/sql/parser/oracle/visitor/statement/impl/OracleDMLStatementSQLVisitor.java:520]: Can we move this logic into StatementContext for processing? The SQLVisitor should be processed according to SQL semantics, and some processing for kernel functions should be placed in StatementContext.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: sql-parser/dialect/oracle/src/main/java/org/apache/shardingsphere/sql/parser/oracle/visitor/statement/impl/OracleDMLStatementSQLVisitor.java (hunk 9 above)
Comment (xinglijun1973) [reply sql-parser/dialect/oracle/src/main/java/org/apache/shardingsphere/sql/parser/oracle/visitor/statement/impl/OracleDMLStatementSQLVisitor.java]: great! I had tried.

File: infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/engine/ProjectionsContextEngine.java [hunk 10]
```diff
@@ -115,32 +114,38 @@ private boolean containsProjection(final OrderByItemSegment orderByItem, final C
...
+            result.add(projection);
         }
         return result;
     }
     
-    private boolean isSameName(final ColumnProjection projection, final String text) {
-        return SQLUtil.getExactlyValue(text).equalsIgnoreCase(projection.getName());
+    private boolean isSameName(final Projection projection, final String text) {
+        String expression = projection.getExpression();
+        return SQLUtil.getExactlyValue(text).equalsIgnoreCase(expression.substring(expression.indexOf('.') + 1));
```
Comment (strongduanmu) [inline infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/engine/ProjectionsContextEngine.java:148]: Why we need `expression.substring(expression.indexOf('.') + 1)`?

File: infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/engine/ProjectionsContextEngine.java [hunk 11]
```diff
@@ -115,32 +114,38 @@ private boolean containsProjection(final OrderByItemSegment orderByItem, final C
     }
     
     private boolean isSameColumn(final Projection projection, final ColumnSegment columnSegment) {
-        Collection<ColumnProjection> columns = getColumnProjections(projection);
+        Collection<Projection> columns = expandProjection(projection);
```
Comment (strongduanmu) [inline infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/select/projection/engine/ProjectionsContextEngine.java:117]: Do you think rename `expandProjection` to `getExpandedProjections` will be better?

File: infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/dml/SelectStatementContext.java [hunk 12]
```diff
@@ -130,6 +162,53 @@ private Map<Integer, SelectStatementContext> createSubqueryContexts(final Shardi
         return result;
     }
     
+    private Map<String, SelectStatementContext> resolveRownumAndContextMap() {
```
Comment (strongduanmu) [inline infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/dml/SelectStatementContext.java:165]: Do you think getRownumSelectStatementContext may be better?

File: infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/dml/SelectStatementContext.java [hunk 13]
```diff
@@ -130,6 +162,53 @@ private Map<Integer, SelectStatementContext> createSubqueryContexts(final Shardi
...
+        Map<SelectStatement, SelectStatementContext> subqueryStatementAndContextMap = new HashMap<>();
+        for (SelectStatementContext subSelectContext : subqueryContexts.values()) {
+            subqueryStatementAndContextMap.put(subSelectContext.getSqlStatement(), subSelectContext);
+        }
+        Map<String, SelectStatementContext> result = new HashMap<>();
+        for (ProjectionSegment proj : getSqlStatement().getProjections().getProjections()) {
```
Comment (strongduanmu) [inline infra/binder/src/main/java/org/apache/shardingsphere/infra/binder/statement/dml/SelectStatementContext.java:171]: Please rename proj to each.

File: infra/binder/src/test/java/org/apache/shardingsphere/infra/binder/statement/dml/SelectStatementContextTest.java [hunk 14]
```diff
@@ -605,4 +613,67 @@ private ProjectionSegment getColumnProjectionSegmentWithoutOwner(final boolean h
...
+        ProjectionsSegment subsubProjections = new ProjectionsSegment(0, 0);
+        ShorthandProjectionSegment shorthandSegment = new ShorthandProjectionSegment(0, 0);
+        shorthandSegment.setOwner(new OwnerSegment(0, 0, new IdentifierValue("a")));
+        subsubProjections.getProjections().add(shorthandSegment);
+        subsubSelectStatement.setProjections(subsubProjections);
+        subsubSelectStatement.setFrom(new SimpleTableSegment(new TableNameSegment(28, 29, new IdentifierValue("a"))));
+        LinkedList<OrderByItemSegment> orderByItems = new LinkedList<OrderByItemSegment>();
+        orderByItems.add(new ColumnOrderByItemSegment(new ColumnSegment(0, 0, new IdentifierValue("name")), OrderDirection.ASC, NullsOrderType.FIRST));
+        subsubSelectStatement.setOrderBy(new OrderBySegment(0, 0, orderByItems));
+        return subsubSelectStatement;
```
Comment (strongduanmu) [inline infra/binder/src/test/java/org/apache/shardingsphere/infra/binder/statement/dml/SelectStatementContextTest.java:669]: Please rename subsubSelectStatement to result.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

**Comment by strongduanmu** [comment]
Since this pr has long time no reponse, I will close it.

**Pull request closed without merging**